Found one feature-wiring issue:

`risingwave_meta_node/test` and `risingwave_meta_service/test` do not compile by themselves. They enable `risingwave_meta/test`, which compiles `src/meta/src/hummock/test_utils.rs`, but that code uses `SstableInfoInner { ..Default::default() }` and the `Default` impl is gated by `risingwave_hummock_sdk/test`.

Repro:

```text
cargo +nightly check -p risingwave_meta_node --features test
```

This fails with:

```text
error[E0277]: the trait bound `SstableInfoInner: std::default::Default` is not satisfied
   --> src/meta/src/hummock/test_utils.rs:183:19
error[E0277]: the trait bound `SstableInfoInner: std::default::Default` is not satisfied
   --> src/meta/src/hummock/test_utils.rs:209:19
```

`risingwave_compaction_test` currently passes because it also depends on `risingwave_storage` with `features = ["test"]`, which indirectly enables `risingwave_hummock_sdk/test`. The new feature itself is still broken for direct users / future package checks. Please propagate `risingwave_hummock_sdk/test` from `risingwave_meta/test`, or explicitly enable it through the new node/service test feature.

--- a/src/meta/Cargo.toml
+++ b/src/meta/Cargo.toml
@@ -8,7 +8,7 @@ license = { workspace = true }
 repository = { workspace = true }
 
 [features]
-test = []
+test = ["risingwave_hummock_sdk/test"]
 failpoints = ["fail/failpoints"]
 
 [dependencies]

--- a/src/meta/src/hummock/manager/versioning.rs
+++ b/src/meta/src/hummock/manager/versioning.rs
@@ -301,8 +301,6 @@ impl HummockManager {
     }
 
     pub async fn may_fill_backward_table_change_logs(&self) -> Result<()> {
-        let mut versioning = self.versioning.write().await;
-
         let is_nonempty_meta_store =
             risingwave_meta_model::hummock_table_change_log::Entity::find()
                 .select_only()
@@ -314,33 +312,33 @@ impl HummockManager {
                 .one(&self.env.meta_store_ref().conn)
                 .await?
                 .is_some();
-        #[expect(deprecated)]
-        if versioning.current_version.table_change_log.is_empty() {
-            tracing::info!("No legacy table change log to migrate.");
-            return Ok(());
-        }
-        let version = Arc::make_mut(&mut versioning.current_version);
-        if is_nonempty_meta_store {
-            tracing::info!("meta store table change log is non-empty.");
+
+        let table_change_logs = {
+            let mut versioning = self.versioning.write().await;
+            #[expect(deprecated)]
+            if versioning.current_version.table_change_log.is_empty() {
+                tracing::info!("No legacy table change log to migrate.");
+                return Ok(());
+            }
+            let version = Arc::make_mut(&mut versioning.current_version);
+            if is_nonempty_meta_store {
+                tracing::info!("meta store table change log is non-empty.");
+                // Clear legacy in-mem state.
+                #[expect(deprecated)]
+                version.table_change_log = HashMap::default();
+                // Either there are no table change logs to commit to the metastore, or the operation has already been completed.
+                return Ok(());
+            }
+
             // Clear legacy in-mem state.
             #[expect(deprecated)]
-            {
-                version.table_change_log = HashMap::default();
-            }
-            // Either there are no table change logs to commit to the metastore, or the operation has already been completed.
-            return Ok(());
-        }
-
-        // Remove table change log from version.
-        #[expect(deprecated)]
-        let table_change_logs = {
-            // Clear legacy in-mem state.
             let logs = std::mem::take(&mut version.table_change_log);
             if logs.values().all(|t| t.is_empty()) {
                 return Ok(());
             }
             logs
         };
+
         // Store table change log in meta store.
         let insert_batch_size = self.env.opts.table_change_log_insert_batch_size as usize;
         let count = {
@@ -376,6 +374,7 @@ impl HummockManager {
         };
         tracing::info!("Migrated {count} table change log to meta store.");
         // Initialize new in-mem state.
+        let mut versioning = self.versioning.write().await;
         versioning.table_change_log = table_change_logs;
         Ok(())
     }

